cmd/monitorCodeGen/main.go

```diff
@@ -0,0 +1,195 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/signalfx/signalfx-agent/internal/monitors"
+	log "github.com/sirupsen/logrus"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"text/template"
+)
+
+var tmpl = `// Code generated by monitor-code-gen. DO NOT EDIT.
+
+package {{.goPackage}}
+
+import "github.com/signalfx/signalfx-agent/internal/monitors"
+
+{{with .groupMetricsMap}}
+const (
+{{- range . }}
+	{{(printf "group.%s" .Name) | formatVariable}} = "{{.Name}}"
+{{- end}}
+)
+{{end}}
+
+var groupSet = map[string]bool {
+{{- range .groupMetricsMap}}
+	{{(printf "group.%s" .Name) | formatVariable}}: true,
+{{- end}}
+}
+
+{{with .metrics}}
+const (
+{{- range .}}
+	{{.Name | formatVariable}} = "{{.Name}}"
+{{- with .Alias}}
+	{{. | formatVariable}} = "{{.}}"
+{{- end}}
+{{- end}}
+)
+{{end}}
+
+var metricSet = map[string]bool {
+{{- range .metrics}}
+	{{.Name | formatVariable}}: true,
+{{- end}}
+}
+
+var includedMetrics = map[string]bool {
+{{- range .metrics}}
+{{- if .Included}}
+	{{.Name | formatVariable}}: true,
+{{- with .Alias}}
+	{{. | formatVariable}}: true,
+{{- end}}
+{{- end}}
+{{- end}}
+}
+
+var groupMetricsMap = map[string][]string {
+{{- range $group, $metrics := .groupMetricsMap}}
+	{{(printf "group.%s" $group) | formatVariable}}: []string {
+		{{- range $metrics}}
+		{{. | formatVariable}},
+		{{- end}}
+	},
+{{- end}}
+}
+
+{{range .monitors}}
+var {{if gt (len $.monitors) 1 -}}
+{{- (printf "%s%s" .MonitorType "monitorMetadata") | formatVariable -}}
+{{else -}}
+monitorMetadata{{end}} = monitors.Metadata{
+	MonitorType: "{{.MonitorType}}",
+	IncludedMetrics: includedMetrics,
+	Metrics: metricSet,
+	MetricsExhaustive: {{.MetricsExhaustive}},
+	Groups: groupSet,
+	GroupMetricsMap: groupMetricsMap,
+	SendAll: {{ .SendAll }},
+}
+{{end}}
+`
+
+// shouldRegenerate determines whether the metadata file needs regenerated based on its existence and timestamps.
+func shouldRegenerate(pkg *monitors.PackageMetadata) (bool, error) {
+	generatedMetadata := filepath.Join(pkg.PackagePath, "generatedMetadata.go")
+
+	var generatorStat os.FileInfo
+	var statMetadataYaml os.FileInfo
+
+	if path, err := os.Executable(); err != nil {
+		return false, err
+	} else if generatorStat, err = os.Stat(path); err != nil {
+		return false, err
+	} else if statMetadataYaml, err = os.Stat(pkg.Path); err != nil {
+		return false, fmt.Errorf("unable to stat %s", pkg.Path)
+	}
+
+	statMetadataGenerated, err := os.Stat(generatedMetadata)
+
+	if err != nil {
+		// generatedMetadata.go does not exist.
+		if os.IsNotExist(err) {
+			return true, nil
+		}
+		return false, fmt.Errorf("unable to stat %s", generatedMetadata)
+	}
+
+	// There's an existing generatedMetadata.go so check timestamps.
+	return statMetadataYaml.ModTime().After(statMetadataGenerated.ModTime()) ||
+		generatorStat.ModTime().After(statMetadataGenerated.ModTime()), nil
+}
+
+func generate() error {
+	pkgs, err := monitors.CollectMetadata("internal/monitors")
```

Comment:
There should be some mechanism somewhere to forcibly rewrite the generated metadata files, even if they exist and have newer timestamps than the source yaml.

Reply:
I can add a specific `clean-code-gen` target. I just do `ack -g 'generatedMetadata.go|template.go' | xargs rm` right now. Or I can add a `--force` flag. Or both. Have a preference?

Reply:
A force flag seems easier to deal with.